<h1>Laptop Sales Analysis</h1>
<h3>Ibra Rizqy Siregar | 221402104</h3>

Tujuan dari analisis ini adalah untuk membuat sebuah dashboard yang membantu tim manajemen dalam perusahaan memahami performa penjualan laptop dengan pertanyaan:
1. Bagaimana tren pendapatan perbulan sepanjang tahun
2. Brand mana yang paling banyak terjual?
3. Bagaimana perbandingan pendapatan penjualan berdasarkan wilayah?
4. Seberapa lama rata-rata stok laptop terjual?
5. Apakah performa laptop memengaruhi kecenderungan customer membeli?

In [182]:
import pandas as pd
df = pd.read_csv('mobile_sales_data.csv')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Product                  50000 non-null  object
 1   Brand                    50000 non-null  object
 2   Product Code             50000 non-null  object
 3   Product Specification    50000 non-null  object
 4   Price                    50000 non-null  int64 
 5   Inward Date              50000 non-null  object
 6   Dispatch Date            50000 non-null  object
 7   Quantity Sold            50000 non-null  int64 
 8   Customer Name            50000 non-null  object
 9   Customer Location        50000 non-null  object
 10  Region                   50000 non-null  object
 11  Core Specification       25017 non-null  object
 12  Processor Specification  50000 non-null  object
 13  RAM                      50000 non-null  object
 14  ROM                      50000 non-nul

Product      Brand Product Code  \
0  Mobile Phone   Motorola     88EB4558   
1        Laptop       Oppo     416DFEEB   
2  Mobile Phone    Samsung     9F975B08   
3        Laptop       Sony     73D2A7CC   
4        Laptop  Microsoft     CCE0B80D   

                               Product Specification   Price Inward Date  \
0  Site candidate activity company there bit insi...   78570  2023-08-02   
1            Beat put care fight affect address his.   44613  2023-10-03   
2           Energy special low seven place audience.  159826  2025-03-19   
3  Friend record hard contain minute we role sea ...   20911  2024-02-06   
4   Program recently feeling save tree hotel people.   69832  2023-08-10   

  Dispatch Date  Quantity Sold  Customer Name Customer Location   Region  \
0    2023-08-03              6   William Hess      South Kelsey  Central   
1    2023-10-06              1    Larry Smith        North Lisa    South   
2    2025-03-20              5  Leah Copeland        South Todd  Central   
3    2024-03-27              1      Dan Kirby   New Jordanmouth    South   
4    2023-09-10              4    Dean Martin          Keithton     East   

  Core Specification Processor Specification   RAM    ROM    SSD  
0                NaN        Snapdragon 7 Gen  12GB  128GB    NaN  
1            Ryzen 5                 Ryzen 5   8GB  512GB  256GB  
2                NaN      MediaTek Dimensity   8GB  256GB    NaN  
3                 i7                      i7  12GB   64GB    2TB  
4                 i7                      i7   8GB  128GB    2TB

Dataset : Mobiles & laptop Sales Data (https://www.kaggle.com/datasets/vinothkannaece/mobiles-and-laptop-sales-data/data)

Dataset yang digunakan berisi 25017 data tentang berbagai informasi dalam penjualan laptop dan handphone. Berikut penjelasan tiap data:
1. Product: Tipe produk apakah handphone atau laptop
2. Brand: Merek laptop/handphone
3. Product Code: Identifier
4. Product Specification: Deskripsi produk
5. Price
6. Inward Date: Tanggal produk masuk ke stok
7. Dispatch Date: Tanggal produk Keluar dari stok
8. Quantity Sokd
9. Customer Name
10. Customer Location
11. Region: Bagian tempat tinggal customer (south, central, west, etc)
12. Core Specification: Spesifikasi processor laptop
13. Core Specification: Spesifikasi processor handphone dan laptop
14. RAM: Random Access Memory, tempat menyimpan memori sementara
15. ROM: Read Only Memory, tempat menyimpan data dalam hardware
16. SSD: Solid State Drive, tempat menyimpan data dalam 

Data Cleaning

In [183]:
# Memisah data untuk produk dan Laptop
df_phones = df[df['Product'] == 'Mobile Phone']
df_phones = df_phones.drop(columns=['Core Specification', 'SSD', 'Product'])

df_phones.to_excel('phones_sales_data.xlsx', index=False)

df_laptops = df[df['Product'] == 'Laptop']
df_laptops.drop(columns=['Product', 'Processor Specification'], inplace=True)
df_laptops.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25017 entries, 1 to 49999
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Brand                  25017 non-null  object
 1   Product Code           25017 non-null  object
 2   Product Specification  25017 non-null  object
 3   Price                  25017 non-null  int64 
 4   Inward Date            25017 non-null  object
 5   Dispatch Date          25017 non-null  object
 6   Quantity Sold          25017 non-null  int64 
 7   Customer Name          25017 non-null  object
 8   Customer Location      25017 non-null  object
 9   Region                 25017 non-null  object
 10  Core Specification     25017 non-null  object
 11  RAM                    25017 non-null  object
 12  ROM                    25017 non-null  object
 13  SSD                    25017 non-null  object
dtypes: int64(2), object(12)
memory usage: 2.9+ MB


C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\1679557296.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops.drop(columns=['Product', 'Processor Specification'], inplace=True)


Membuat kolom baru

In [184]:
# Membuat Total Revenue dan Inventory Time
df_laptops['Total Revenue'] = df['Price'] * df['Quantity Sold']
df['Inward Date'] = pd.to_datetime(df_laptops['Inward Date'])
df['Dispatch Date'] = pd.to_datetime(df_laptops['Dispatch Date'])
df_laptops['Inventory Time'] =  df['Dispatch Date'] - df['Inward Date']


C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\1258888416.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops['Total Revenue'] = df['Price'] * df['Quantity Sold']
C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\1258888416.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops['Inventory Time'] =  df['Dispatch Date'] - df['Inward Date']


Training untuk kolom Performance Rating

In [185]:
print(df_laptops['Core Specification'].unique())
print(df_laptops['RAM'].unique())
print(df_laptops['SSD'].unique())

['Ryzen 5' 'i7' 'i5' 'Ryzen 3' 'i3' 'i9' 'Ryzen 7' 'Ryzen 9']
['8GB' '12GB' '16GB' '6GB' '32GB' '4GB']
['256GB' '2TB' '1TB' '512GB']


In [186]:
#Mapping core specifications, RAM, and SSD ke numerik
core_spec_mapping = {
    'i3': 1,
    'i5': 2,
    'i7': 3,
    'i9': 4,
    'Ryzen 3': 1,
    'Ryzen 5': 2,
    'Ryzen 7': 3,
    'Ryzen 9': 4
}
ram_mapping = {
    '4GB': 4,
    '6GB': 6,
    '8GB': 8,
    '12GB': 12,
    '16GB': 16,
    '32GB': 32
}
ssd_mapping = {
    '256GB': 256,
    '512GB': 512,
    '1TB': 1024,
    '2TB': 2048
}
df_laptops['Core Specification'] = df_laptops['Core Specification'].map(core_spec_mapping)
df_laptops['RAM'] = df_laptops['RAM'].map(ram_mapping)
df_laptops['SSD'] = df_laptops['SSD'].map(ssd_mapping)

C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\3182941349.py:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops['Core Specification'] = df_laptops['Core Specification'].map(core_spec_mapping)
C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\3182941349.py:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops['RAM'] = df_laptops['RAM'].map(ram_mapping)
C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\3182941349.py:28: SettingWithCopyWarning: 
A value is trying to be set on a copy of a sl

In [ ]:
# Training dengan KMeans
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler

features = df_laptops[['Core Specification', 'RAM', 'SSD']]
scaler = StandardScaler()
features_scaled = scaler.fit_transform(features)
kmeans = KMeans(n_clusters=3, random_state=42)
kmeans.fit(features_scaled)

labels_mapping = {0: 'Low', 1: 'Medium', 2: 'High'}
df_laptops['Performance Rating'] = kmeans.labels_
df_laptops['Performance Rating'] = df_laptops['Performance Rating'].map(labels_mapping)


C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\1482368974.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops['Performance Rating'] = kmeans.labels_
C:\Users\rizqy\AppData\Local\Temp\ipykernel_21352\1482368974.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_laptops['Performance Rating'] = df_laptops['Performance Rating'].map(labels_mapping)


In [188]:
df_laptops.info()
df_laptops.sample(5)
df_laptops

<class 'pandas.core.frame.DataFrame'>
Index: 25017 entries, 1 to 49999
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype          
---  ------                 --------------  -----          
 0   Brand                  25017 non-null  object         
 1   Product Code           25017 non-null  object         
 2   Product Specification  25017 non-null  object         
 3   Price                  25017 non-null  int64          
 4   Inward Date            25017 non-null  object         
 5   Dispatch Date          25017 non-null  object         
 6   Quantity Sold          25017 non-null  int64          
 7   Customer Name          25017 non-null  object         
 8   Customer Location      25017 non-null  object         
 9   Region                 25017 non-null  object         
 10  Core Specification     25017 non-null  int64          
 11  RAM                    25017 non-null  int64          
 12  ROM                    25017 non-null  object      

Brand Product Code  \
1           Oppo     416DFEEB   
3           Sony     73D2A7CC   
4      Microsoft     CCE0B80D   
5             HP     10A5C53D   
7          Apple     D58FF785   
...          ...          ...   
49994     Huawei     D1AB0532   
49995   Motorola     8C187637   
49996         HP     0DA9E4D3   
49997    Toshiba     CAD89A93   
49999       Asus     9C3FC57C   

                                   Product Specification   Price Inward Date  \
1                Beat put care fight affect address his.   44613  2023-10-03   
3      Friend record hard contain minute we role sea ...   20911  2024-02-06   
4       Program recently feeling save tree hotel people.   69832  2023-08-10   
5      Recognize happen middle clear simple player po...  190474  2025-03-19   
7      Reality sea weight thing newspaper good next g...   15041  2023-10-29   
...                                                  ...     ...         ...   
49994  Community walk as form line century state seco...  131689  2024-08-15   
49995            Pick build kind teach mean each camera.  170385  2024-03-02   
49996       Outside policy treat he opportunity pattern.  193133  2025-01-19   
49997  Test floor tax together out adult discover own...   66502  2024-01-16   
49999      Role ahead common listen idea few blood half.   86483  2023-08-29   

      Dispatch Date  Quantity Sold   Customer Name      Customer Location  \
1        2023-10-06              1     Larry Smith             North Lisa   
3        2024-03-27              1       Dan Kirby        New Jordanmouth   
4        2023-09-10              4     Dean Martin               Keithton   
5        2025-05-08              1    Monica Berry           West Heather   
7        2023-12-20              1  Danielle Smith           North Travis   
...             ...            ...             ...                    ...   
49994    2024-08-24              1     Paula Kelly           West Michael   
49995    2024-04-11              6      Sara Gomez             Sylviatown   
49996    2025-01-20              4   Raymond Clark             Chrismouth   
49997    2024-03-05              6   Joseph Martin               Huntfurt   
49999    2023-10-26              7  Robert Sampson  West Elizabethchester   

        Region  Core Specification  RAM    ROM   SSD  Total Revenue  \
1        South                   2    8  512GB   256          44613   
3        South                   3   12   64GB  2048          20911   
4         East                   3    8  128GB  2048         279328   
5      Central                   3   16   64GB   256         190474   
7        South                   3   12  128GB   256          15041   
...        ...                 ...  ...    ...   ...            ...   
49994  Central                   4    4  512GB   512         131689   
49995    North                   1   16  512GB  2048        1022310   
49996     West                   3    8    1TB   512         772532   
49997    North                   2    4   64GB  2048         399012   
49999     East                   3   12  128GB   256         605381   

      Inventory Time Performance Rating  
1             3 days             Medium  
3            50 days                Low  
4            31 days                Low  
5            50 days               High  
7            52 days               High  
...              ...                ...  
49994         9 days               High  
49995        40 days                Low  
49996         1 days             Medium  
49997        49 days                Low  
49999        58 days               High  

[25017 rows x 17 columns]

In [ ]:
#Simpan DataFrame ke file Excel
df_laptops.to_excel('laptops_sales_data.xlsx', index=False)
